File: LeetCode/Algorithm/C++/15. 01 Matrix.cpp
/*
01 Matrix

Given an m x n binary matrix mat, return the distance of the nearest 0 for each cell.
The distance between two adjacent cells is 1.

Example 1:

Input: mat = [[0,0,0],[0,1,0],[0,0,0]]
Output: [[0,0,0],[0,1,0],[0,0,0]]

Example 2:

Input: mat = [[0,0,0],[0,1,0],[1,1,1]]
Output: [[0,0,0],[0,1,0],[1,2,1]]

Constraints:

    m == mat.length
    n == mat[i].length
    1 <= m, n <= 104
    1 <= m * n <= 104
    mat[i][j] is either 0 or 1.
    There is at least one 0 in mat.
*/
#include <bits/stdc++.h>
using namespace std;
bool isValid(int row, int col, int rowsize, int colsize) {
    if(row < 0 || row >= rowsize || col < 0 || col >= colsize)
        return false;
    return true;
}
vector<vector<int>> updateMatrix(vector<vector<int>>& mat) {
    vector<vector<int>> matrix(mat.size(),vector<int>(mat[0].size(),-1));
    queue<pair<int,int>> q;
    for(int row = 0; row < mat.size(); row++) {
        for(int col = 0; col < mat[0].size(); col++) {
            if(mat[row][col] == 0) {
                matrix[row][col] = 0;
                q.push({row,col});
            }
        }
    }
    while(!q.empty()) {
        int row = q.front().first;
        int col = q.front().second;
        q.pop();
        if(isValid(row-1,col,matrix.size(),matrix[0].size()) && matrix[row-1][col] == -1) {
            matrix[row-1][col] = 1+matrix[row][col];
            q.push({row-1,col});
        }
        if(isValid(row+1,col,matrix.size(),matrix[0].size()) && matrix[row+1][col] == -1) {
            matrix[row+1][col] = 1+matrix[row][col];
            q.push({row+1,col});
        }
        if(isValid(row,col-1,matrix.size(),matrix[0].size()) && matrix[row][col-1] == -1) {
            matrix[row][col-1] = 1+matrix[row][col];
            q.push({row,col-1});
        }
        if(isValid(row,col+1,matrix.size(),matrix[0].size()) && matrix[row][col+1] == -1) {
            matrix[row][col+1] = 1+matrix[row][col];
            q.push({row,col+1});
        }
    }
    return matrix;
}
int main() {
    int m,n;
    cin >> m >> n;
    vector<vector<int>> mat(m,vector<int>(n,0));
    for(int row = 0; row < m; row++)
        for(int col = 0; col < n; col++)
            cin >> mat[row][col];
    cout << endl;
    vector<vector<int>> matrix = updateMatrix(mat);
    for(auto v: matrix) {
        for(auto x: v)
            cout << x << " ";
        cout << endl;
    }
}

File: LeetCode/Algorithm/Python/15. 01 Matrix.py
'''
01 Matrix

Given an m x n binary matrix mat, return the distance of the nearest 0 for each cell.
The distance between two adjacent cells is 1.

Example 1:

Input: mat = [[0,0,0],[0,1,0],[0,0,0]]
Output: [[0,0,0],[0,1,0],[0,0,0]]

Example 2:

Input: mat = [[0,0,0],[0,1,0],[1,1,1]]
Output: [[0,0,0],[0,1,0],[1,2,1]]

Constraints:

    m == mat.length
    n == mat[i].length
    1 <= m, n <= 104
    1 <= m * n <= 104
    mat[i][j] is either 0 or 1.
    There is at least one 0 in mat.
'''
def isValid(r: int, c: int, rsize: int, csize: int):
    if r < 0 or r >= rsize or c < 0 or c >= csize:
        return False
    return True
def updateMatrix(mat: list[list[int]]) -> list[list[int]]:
    q = []
    ans = [[-1 for c in range(len(mat[0]))] for r in range(len(mat))]
    for r in range(len(mat)):
        for c in range(len(mat[0])):
            if mat[r][c] == 0:
                q.append((r,c))
                ans[r][c] = 0
    while len(q) > 0:
        r,c = q[0][0],q[0][1]
        q.pop(0)
        if isValid(r-1,c,len(mat),len(mat[0])) and ans[r-1][c] == -1:
            ans[r-1][c] = 1+ans[r][c]
            q.append((r-1,c))
        if isValid(r+1,c,len(mat),len(mat[0])) and ans[r+1][c] == -1:
            ans[r+1][c] = 1+ans[r][c]
            q.append((r+1,c))
        if isValid(r,c-1,len(mat),len(mat[0])) and ans[r][c-1] == -1:
            ans[r][c-1] = 1+ans[r][c]
            q.append((r,c-1))
        if isValid(r,c+1,len(mat),len(mat[0])) and ans[r][c+1] == -1:
            ans[r][c+1] = 1+ans[r][c]
            q.append((r,c+1))
    return ans

if __name__ == "__main__":
    n,m = map(int,input().split())
    mat = [[int(x) for x in input().split()] for r in range(n)]
    ans = updateMatrix(mat)
    print(ans)